wrapper/src/main/java/software/amazon/jdbc/util/CacheMap.java

```diff
@@ -21,51 +21,106 @@
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
 
 public class CacheMap<K, V> {
 
-  private final Map<K, CacheItem<V>> cache = new ConcurrentHashMap<>();
-  private final long cleanupIntervalNanos = TimeUnit.MINUTES.toNanos(10);
+  private final Map<K, CacheItem> cache = new ConcurrentHashMap<>();
+  private long cleanupIntervalNanos = TimeUnit.MINUTES.toNanos(10);
   private final AtomicLong cleanupTimeNanos = new AtomicLong(System.nanoTime() + cleanupIntervalNanos);
+  private final IsItemValidFunc<V> isItemValidFunc;
+  private final ItemDisposalFunc<V> itemDisposalFunc;
 
   public CacheMap() {
+    this.isItemValidFunc = null;
+    this.itemDisposalFunc = null;
+  }
+
+  public CacheMap(
+      final IsItemValidFunc<V> isItemValidFunc,
+      final ItemDisposalFunc<V> itemDisposalFunc) {
+    this.isItemValidFunc = isItemValidFunc;
+    this.itemDisposalFunc = itemDisposalFunc;
   }
 
   public V get(final K key) {
-    final CacheItem<V> cacheItem = cache.computeIfPresent(key, (kk, vv) -> vv.isExpired() ? null : vv);
+    removeIfExpired(key);
+    final CacheItem cacheItem = cache.get(key);
     return cacheItem == null ? null : cacheItem.item;
   }
 
   public V get(final K key, final V defaultItemValue, final long itemExpirationNano) {
```

Comment:
do we need both this and computeIfAbsent? They are quite similar. Differences: computeIfAbsent calls cleanup and extends the expiration